TRD PID Study

KangkanGoswami opened the pull request:
This change contains the extra tables for TRD PID Study.

KangkanGoswami committed:
Update AODProducerWorkflowSpec.h

KangkanGoswami committed:
Update AODProducerWorkflowSpec.cxx

KangkanGoswami committed:
Update aod-producer-workflow.cxx

KangkanGoswami committed:
Update AnalysisDataModel.h

alibuild committed:
Please consider the following formatting changes

KangkanGoswami committed:
Merge pull request #1 from alibuild/alibot-cleanup-14683

Please consider the following formatting changes to #14683

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 47a07e761ff9926ca61878421cfe68c11d79b463 at 2025-09-24 14:16:
```
## sw/BUILD/O2-latest/log
/sw/SOURCES/O2/14683-slc9_x86-64/0/Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1991:47: error: redeclaration of 'o2::framework::Produces<o2::soa::Table<o2::aod::Hash<4125400158>, o2::aod::Hash<3317288390>, o2::aod::Hash<2286545062> > > ft0ExtraCursor'
/sw/SOURCES/O2/14683-slc9_x86-64/0/Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1992:47: error: redeclaration of 'o2::framework::Produces<o2::soa::Table<o2::aod::Hash<2440583501>, o2::aod::Hash<1097011421>, o2::aod::Hash<2286545062> > > fddExtraCursor'
/sw/SOURCES/O2/14683-slc9_x86-64/0/Detectors/AOD/src/AODProducerWorkflowSpec.cxx:1993:48: error: redeclaration of 'o2::framework::Produces<o2::soa::Table<o2::aod::Hash<3518865614>, o2::aod::Hash<898762686>, o2::aod::Hash<2286545062> > > fv0aExtraCursor'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14683/47a07e761ff9926ca61878421cfe68c11d79b463/build_O2_fullCI_slc9/pretty.html).

f3sch commented on Detectors/AOD/src/AODProducerWorkflowSpec.cxx:
```diff
@@ -1910,6 +1987,20 @@ void AODProducerWorkflowDPL::run(ProcessingContext& pc)
   auto originCursor = createTableCursor<o2::aod::Origins>(pc);
 
+  /// Extra tables
+  o2::framework::Produces<o2::aod::FT0sExtra> ft0ExtraCursor;
+  o2::framework::Produces<o2::aod::FDDsExtra> fddExtraCursor;
+  o2::framework::Produces<o2::aod::FV0AsExtra> fv0aExtraCursor;
```
Hi @KangkanGoswami, the build error is genuine. Apparently, my patches need to be updated. Removing the previous definitions of `ft0ExtraCursor`, `fddExtraCursor` and `fv0aExtraCursor` before `// Extra tables` suffices.

KangkanGoswami replied:
Hi @f3sch, thank you so much for your suggestion. I have now removed these three lines, which should take care of the error.

f3sch replied:
No, this removed the wrong lines and broke effectively the option `---enable-FIT-extra`.
The lines I marked should be kept, while removing the lines `  auto fv0aExtraCursor = createTableCursor<o2::aod::FV0AsExtra>(pc);`, e.g. lines 1967&1969&1970.

KangkanGoswami replied:
Oh, my bad. Now, I think this should work.

KangkanGoswami committed:
Update AODProducerWorkflowSpec.cxx

I have now removed ft0ExtraCursor, fddExtraCursor, and fv0aExtraCursor. This should, in principle, resolve the issue of re-definition.

KangkanGoswami committed:
Update AODProducerWorkflowSpec.cxx

Corrected re-definition error.

alibuild committed:
Please consider the following formatting changes

KangkanGoswami committed:
Merge pull request #2 from alibuild/alibot-cleanup-14683

Please consider the following formatting changes to #14683

f3sch commented:
@jgrosseo, could you please approve for the missing security check, so at least the CI can run, thanks!

sawenzel reviewed (changes requested):
This is a major addition to the AOD writer and data model. To continue here, I think this should first of all be presented and discussed in the Analysis meeting (if it hasn't happened yet) or be supported by Physics Coordination. It would also be good to know if this is supposed to be included by default or only in very specific studies.

For now a few comments from my side are:

(a) please try to give a more comprehensive message to the PR. Just saying "This is for the TRD PID study" is a bit too short. Good common practice is "This PR provides new (optional) tables in the AOD that are needed for...; This has been discussed in this **JIRA ticket** or in these **slides**" 

(b) please try to cleanup this PR. It should only be 1 clean commit

sawenzel commented on Detectors/AOD/src/AODProducerWorkflowSpec.cxx:
```diff
@@ -390,6 +392,79 @@ void AODProducerWorkflowDPL::addToTracksQATable(TracksQACursorType& tracksQACurs
+template <typename TRDsExtraCursorType>
+void AODProducerWorkflowDPL::addToTRDsExtra(const o2::globaltracking::RecoContainer& recoData, TRDsExtraCursorType& trdExtraCursor, const GIndex& trkIdx, int trkTableIdx)
+{
+  static int q0s[6] = {-1}, q1s[6] = {-1}, q2s[6] = {-1};
```
This is a lot of new detector specific code. Could this not be done in TRD reco for better logic isolation? Also, I would be interested to know the CPU impact on the overall AOD processing.

KangkanGoswami commented:
Hi @sawenzel, my apologies for such a messy PR. I presented these changes in the weekly TRD meeting. To clean up this PR, I would close the current one and create a new PR as soon as possible.

sawenzel commented:
> Hi @sawenzel, my apologies for such a messy PR. I presented these changes in the weekly TRD meeting. To clean up this PR, I would close the current one and create a new PR as soon as possible.

Thanks but you don't need to close it. You can just squash locally + make a good commit message and force-push to the same PR.

KangkanGoswami committed:
Merge branch 'AliceO2Group:dev' into dev

KangkanGoswami committed:
Merge branch 'AliceO2Group:dev' into dev

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 111e983f8585ecc073ca6cbe5b21191047e1bac7 at 2025-10-08 05:51:
```
No log files found
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14683/111e983f8585ecc073ca6cbe5b21191047e1bac7/build_O2_fullCI_slc9/pretty.html).

KangkanGoswami committed:
Merge branch 'AliceO2Group:dev' into dev

KangkanGoswami committed:
Merge branch 'AliceO2Group:dev' into dev

KangkanGoswami commented:
This commit requires some cleanup and a clearer description. Due to this, I am closing this commit and starting a new one with a few modifications.
